File: node/sql_agent.js
const fs = require('fs');
const path = require('path');
const sqlite3 = require('sqlite3').verbose();
const axios = require('axios');
require('dotenv').config();

const apiKey = process.env.GROQ_API_KEY;

const connectDb = (dbPath) => {
  const db = new sqlite3.Database(dbPath, (err) => {
    if (err) {
      console.error('Error opening database:', err.message);
    } else {
      console.log('Connected to the SQLite database.');
    }
  });
  return db;
};

const executeQuery = (db, query) => {
  return new Promise((resolve, reject) => {
    db.all(query, [], (err, rows) => {
      if (err) {
        reject(err);
      } else {
        resolve(rows);
      }
    });
  });
};

const listColumns = (db, tableName) => {
  return new Promise((resolve, reject) => {
    db.all(`PRAGMA table_info(${tableName})`, [], (err, rows) => {
      if (err) {
        reject(err);
      } else {
        if (rows.length === 0) {
          console.error(`No columns found for table: ${tableName}`);
        }
        const columns = rows.map((info) => info.name);
        resolve(columns);
      }
    });
  });
};

const validateSqlQuery = (query, columns) => {
  for (const column of columns) {
    if (query.includes(column)) {
      return query;
    }
  }
  throw new Error('The generated SQL query contains non-existent columns.');
};

const questionToSql = async (question, columns) => {
  const prompt = {
    model: 'llama3-8b-8192',
    messages: [
      { role: 'system', content: 'You are a helpful assistant.' },
      { role: 'user', content: `Convert the following natural language question into a SQL query using the columns: ${columns.join(', ')} in the Customers table.\n\nQuestion: ${question}` },
    ],
    max_tokens: 150,
    temperature: 0.5,
  };
  const headers = {
    Authorization: `Bearer ${apiKey}`,
    'Content-Type': 'application/json',
  };
  try {
    const response = await axios.post('https://api.groq.com/openai/v1/chat/completions', prompt, { headers });
    if (response.data.choices) {
      let sqlQuery = response.data.choices[0].message.content;
      sqlQuery = sqlQuery.split('```sql')[1].split('```')[0].trim();
      sqlQuery = validateSqlQuery(sqlQuery, columns);
      return sqlQuery;
    } else {
      throw new Error('Unexpected response format from Groq API');
    }
  } catch (error) {
    throw new Error(`Error: ${error.response ? error.response.data.error.message : error.message}`);
  }
};

const cleanUpResponse = (response) => {
  let cleanedResponse = response.replace('Here is the natural language answer:\n\n', '').trim();
  cleanedResponse = cleanedResponse.replace('Here is the list of employees from the company:\n\n', '').trim();
  return cleanedResponse;
};

const resultToAnswer = async (results) => {
  const prompt = {
    model: 'llama3-8b-8192',
    messages: [
      { role: 'system', content: 'You are a helpful assistant.' },
      { role: 'user', content: `Convert the following SQL query result into a natural language answer:\n\nResults: ${JSON.stringify(results)}` },
    ],
    max_tokens: 150,
    temperature: 0.5,
  };
  const headers = {
    Authorization: `Bearer ${apiKey}`,
    'Content-Type': 'application/json',
  };
  try {
    const response = await axios.post('https://api.groq.com/openai/v1/chat/completions', prompt, { headers });
    if (response.data.choices) {
      let answer = response.data.choices[0].message.content;
      answer = cleanUpResponse(answer);
      return answer;
    } else {
      throw new Error('Unexpected response format from Groq API');
    }
  } catch (error) {
    throw new Error(`Error: ${error.response ? error.response.data.error.message : error.message}`);
  }
};

const main = async () => {
  const dbPath = path.join(__dirname, 'northwind.db'); 
  const db = connectDb(dbPath);

  const columns = await listColumns(db, 'Customers');

  console.log('AI SQL Agent is ready. Type "exit" to quit.');
  const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  readline.on('line', async (question) => {
    if (question.toLowerCase() === 'exit') {
      readline.close();
      db.close();
      return;
    }

    try {
      const sqlQuery = await questionToSql(question, columns);
      const results = await executeQuery(db, sqlQuery);
      const answer = await resultToAnswer(results);
      console.log(`AI: ${answer}`);
    } catch (error) {
      console.error(error.message);
    }
  });
};

main();

File: python/sql_agent.py
import os
import sqlite3
import requests
from dotenv import load_dotenv

load_dotenv()
api_key = os.getenv('GROQ_API_KEY')

def connect_db(db_path):
    conn = sqlite3.connect(db_path)
    return conn

def execute_query(conn, query):
    cursor = conn.cursor()
    cursor.execute(query)
    results = cursor.fetchall()
    cursor.close()
    return results

def list_columns(conn, table_name):
    cursor = conn.cursor()
    cursor.execute(f"PRAGMA table_info({table_name})")
    columns = [info[1] for info in cursor.fetchall()]
    cursor.close()
    return columns

def validate_sql_query(query, columns):
    for column in columns:
        if column in query:
            return query
    raise ValueError("The generated SQL query contains non-existent columns.")

def question_to_sql(question, columns):
    prompt = {
        "model": "llama3-8b-8192",
        "messages": [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": f"Convert the following natural language question into a SQL query using the columns: {', '.join(columns)}.\n\nQuestion: {question}"}
        ],
        "max_tokens": 150,
        "temperature": 0.5
    }
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }
    response = requests.post(
        "https://api.groq.com/openai/v1/chat/completions",
        headers=headers,
        json=prompt
    )
    response_data = response.json()
    if 'choices' in response_data:
        sql_query = response_data['choices'][0]['message']['content']
        sql_query = sql_query.split('```sql')[1].split('```')[0].strip()
        sql_query = validate_sql_query(sql_query, columns)
        return sql_query
    else:
        raise ValueError("Unexpected response format from Groq API")

def result_to_answer(results):
    prompt = {
        "model": "llama3-8b-8192",
        "messages": [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": f"Convert the following SQL query result into a natural language answer:\n\nResults: {results}"}
        ],
        "max_tokens": 150,
        "temperature": 0.5
    }
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }
    response = requests.post(
        "https://api.groq.com/openai/v1/chat/completions", 
        headers=headers,
        json=prompt
    )
    response_data = response.json()
    if 'choices' in response_data:
        answer = response_data['choices'][0]['message']['content']
        answer = answer.replace("Here is the natural language answer:\n\n", "").strip()
        return answer
    else:
        raise ValueError("Unexpected response format from Groq API")

def main():
    db_path = 'northwind.db' 
    conn = connect_db(db_path)

    columns = list_columns(conn, 'Customers')

    print("AI SQL Agent is ready. Type 'exit' to quit.")
    while True:
        question = input("You: ")
        if question.lower() == 'exit':
            break

        try:
            sql_query = question_to_sql(question, columns)
            results = execute_query(conn, sql_query)
            answer = result_to_answer(results)
            print(f"AI: {answer}")
        except Exception as e:
            print(f"Error: {str(e)}")

    conn.close()

if __name__ == '__main__':
    main()
